### PR TITLE
🔙 Improved NACKs

### LINKED ISSUE #129: Test packet handling, especially keyframes and lost packets
Meant to add a functional test covering NACKs as part of #122 but ran out of time.

Also a related regression happened around keyframe generation (#128), would be great to also test that after sending keyframe packets they are returned from the `GetKeyframe()` call.

### PR DESCRIPTION
**Pull request by danstiner**
### Checklist
- [x] Test on a production server (with and without NACKs enabled) (live on `do-sfo3-ingest1.ksfo.live.glimesh.tv` with NACKs enabled)
- [x] Doc comments
- [x] Cleanup code
- [X] Overnight stability test
- [X] Cleanup tests (remove randomized tests, they should be reviewed separately)
- [x] Test if MAX_OUTSTANDING_NACKS should be increased

### Background
Shoutout to Hayden for writing the original NACK (negative acknowledgement of missing packets) feature. It mostly worked but had some issues with sequence number wrap-around (to be fair, I introduced a number of bugs when I touched the NACK code recently). If I remember correctly streams could break if there was significant packet loss (and would not recover when the packet loss went away). This left the feature unusable and it was turned off.

The NACK feature has always been difficult to test and is tied together with the keyframe packet collection. This PR is my second take at fixing the NACK feature by keeping the core ideas of the old code while starting to pull the logic into separate classes for clarity and test-ability.

### Goals
- Bring NACK code under unit tests
- Mitigate up to 1% packet loss between streamer <-> ingest when packet arrival jitter is less than 20ms
- Handle higher packet loss gracefully, stream should recover within a couple seconds (one keyframe interval) if packet loss drops down back under 1%

I think meeting those goals will yield a "good enough" solution, but if I had unlimited time I'd aim for something frame-aware that can do things like prioritize keyframe packet loss and estimate when a packet is "late" based on the frame timestamps, jitter estimates, etc. Those are all things the client does with it's "playout buffer" data structure, we may be able to take inspiration from that and adjust the ideas to be relevant on the server.

### Status
Tested locally and as of 0d23274 it seems effective and stable. Overnight stability tests show it is capable of reducing 1% packet loss to about 0.001%, which is exactly (1% * 1%) as expected (re-transmitted packets can still be lost, the 1% loss rate in this local test also applies to them and this code does not try to re-transmit a second time). This should be the difference between bad frames every couple seconds to more like once a minute.

### Details

There are roughly two parts to the tracking and NACK process:

First, when a packet arrives it is both fed to the sequence tracker instance for the source and relayed immediately to all client sessions. This follows the previous behavior of this code, no latency is added to packet relaying. The sequence tracker checks if the sequence number is a re-transmit due to a prior NACK. If so, that NACK is marked as received and it does nothing else. If this is a new packet, it is added to an internal buffer, in sequence order.

Second, when it's time to send NACKs, we iterate through the buffer of received packets and look for gaps in the sequence numbers. Any gaps are marked as "missing" packets and NACKs are sent for the missing packets. As mentioned above, when NACK'd packets are re-transmitted and received they are removed from the missing list. If we never get the re-transmit, eventually we time out the NACK and stop tracking the missing packet.

That's the jist of it, but there are some additional complexities:
- There is an allowance for re-ordered RTP packets, just like the old logic we do not check the highest 16 sequence numbers in the packet buffer for gaps/missing packets
  - aka a sliding window based on sequence number
  - Note this allowance is *not* guaranteed to cover the last 16 packets received, it is sequence number based. e.g. if we receive packet with seq=16 and then the next packet has seq=32, then we will check for gaps in packets with seq<=16. Sequence numbers 17-31 will not be considered missing as they are still in the "re-order window", at least until yet newer packets arrive and slide the window forward.
- There is a [watermark](https://en.wikipedia.org/wiki/Watermark_(data_synchronization)) (checkForMissingWatermark) tracking the highest sequence number that has been checked for gaps. This ensures we do not count the same gap as missing twice.
  - Watermark is not a great name for this, open to alternatives
- There is a limit on the number of in-flight/outstanding NACKs to prevent excessive re-transmit bandwidth usage. If we are dropping packets to the ingest due to bandwidth limits, it will not be good to use even more bandwidth! (new in this PR I think)

Also, snuck in a change to the pending/current keyframe packet collection logic in FtlMediaConnection. It now checks for sequential packets without gaps and that the final packet has a marker bit before it considers a keyframe valid. This should prevent most cases of corrupt frames being used at stream thumbnails.

Closes #129

### Potential Future Improvements

- Better name for the SequenceTracker class?
- Combine ExtendedSequenceCounter with SequenceTracker, their logic is substantially similar
- Look at if extended sequence numbers are needed outside of SequenceTracker, maybe can limit use / get rid of that concept entirely
- Bring back NACK bitmasks to reduce packets being sent / only send NACKs periodically in batches vs processing them on every packet received (observed testing on a wifi network shows packets are often lost in bursts of ~3-20 sequential packets)
  - Note that for simplicity this code runs the check for gaps/missing packets and sends any ready NACKs every time a new packet is received. This could lead to sliding the "re-order window" forward by one sequence number at a time, meaning even if a burst of sequence packets was lost, the window would only slide to reveal one of them at a time and thus singular NACKs would still be sent even with bitmask support. I don't have a great solution for this, probably the check for missing/sending NACKs should be on some kind of a timer instead of done per-packet.
  - [ ] TODO open issue for this
- Group packets in buffer into frames by their RTP timestamps (allows e.g. prioritize NACKs of keyframes, if refactored enough this could replace the logic in FtlMediaConnection that captures keyframe packets)
- Redefine timeouts to be based on RTP timestamps and expected arrival time (may be for a future PR as this would require extra logic for jitter/arrival time estimation)
- Estimate expected arrival time for packets, to allow better time-based NACKing. Right now we have to wait until a newer packet arrives before we can calculate packets are missing and send NACKs.  (more detail in below comment)

### CODE REVIEW

**Comment by danstiner** [comment]
Thanks for deploying @clone1018, excited to see this out in the wild! I tested a bit and can confirm it appears to be working.

For the fun of it I did some packet capturing. From the viewer side, the following is an example of where Seq 62850-62852 for SSRC 0x5EFD5FAE did not arrive initially:
```

    No.           UTC         Protocol   Length                                     Info                                    
 --------- ----------------- ---------- -------- -------------------------------------------------------------------------- 
  1029968   22:23:31.353413   RTP          1462   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62848, Time=1046226000        
  1029969   22:23:31.353438   RTP          1437   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62849, Time=1046226000, Mark  

  1029989   22:23:31.473773   RTP          1462   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62853, Time=1046229000        
  1029990   22:23:31.473806   RTP          1462   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62854, Time=1046229000        
```

Then from the streamer's side, OBS logs show we got the NACK from `do-sfo3-ingest1` and it re-transmitted those three missing packets.
Side note: Client/server time are both in UTC but not quite in sync, that's pretty normal and does not hurt anything.
```
22:23:31.295: [3] [12677] resent sn 62850, request delay was 196 ms, was part of iframe? 0
22:23:31.295: [3] [12677] resent sn 62851, request delay was 197 ms, was part of iframe? 0
22:23:31.295: [3] [12677] resent sn 62852, request delay was 197 ms, was part of iframe? 0
```

Then a bit later on the viewer side we see the re-transmitted packets arrive just after Seq 62929, about 190ms after they should have arrived, which matches the OBS logs. In this case it is likely the re-transmit arrived in time and the user noticed no lost frames, but that depends on the playout delay Chrome has chosen. (Note applying my playout-delay patch is one way to ensure there is a sufficient delay for re-transmits like this to come through, especially in low packet loss situations Chrome could lower the playout delay too aggressively for our re-transmits to arrive in time)

```
    No.           UTC         Protocol   Length                                     Info                                    
 --------- ----------------- ---------- -------- -------------------------------------------------------------------------- 
  1030150   22:23:31.539324   RTP          1462   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62928, Time=1046245500        
  1030151   22:23:31.539324   RTP           539   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62929, Time=1046245500, Mark  

  1030163   22:23:31.541572   RTP          1462   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62850, Time=1046227500        
  1030164   22:23:31.541572   RTP          1462   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62851, Time=1046227500        
  1030165   22:23:31.541572   RTP           724   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62852, Time=1046227500, Mark  
 
  1030175   22:23:31.568032   RTP          1462   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62930, Time=1046247000        
  1030176   22:23:31.568067   RTP          1462   PT=DynamicRTP-Type-96, SSRC=0x5EFD5FAE, Seq=62931, Time=1046247000        
```

A future improvement would be to lower the re-transmit delay. Especially because I see some re-transmit requests with delay in the 500-750ms range, which probably arrive too late to be of any use because the video has already advanced past that frame. I suspect this is especially a problem for audio packets which are sent at a low rate. The current code waits until a certain number of newer packets have been seen before considering a packet lost. This allows for out-of-order packet arrival, but can greatly delay the re-transmit if not enough newer packets arrive. Instead, as mentioned above, a time-based calculation of when a packet is missing/late could help. Ideally, allowing for a 30ms reorder window plus my round-trip time of 27ms should result in a re-transmit request delay of around 57ms. Realistically that may not be possible but a delay under 100ms seems plausible.

tl;dr: It works pretty well, but sometimes re-transmitted packets arrive too late and video/audio skips still can happen. Applying a playout-delay of 400ms+ would likely help mitigate the late arriving re-transmits until this code can be further improved.

**Comment by clone1018** [comment]
Sounds like a good approach to me. It is worth mentioning that smashbets has notoriously finicky internet when it comes to FTL, even having documented issues going back to the Mixer days. It's a good test, but will sometimes have significantly more packet loss than this case can cover. 

I have no problem deploying the playout-delay as well, but a combined roll out seems easiest from a downtime perspective. Do you have any tests you want to run on production specifically for playout-delay? We could always deploy it to one server for now, and then a combined rollout later.

**Comment by danstiner** [comment]
Yeah I worked with SmashBets some on their connection issues, but it looks like they are more stable at a higher resolution/bitrate than they could do before so that's good news! Even if it isn't perfect.

Sounds good on deployment, deploying playout-delay to just one edge to start with would be good. I opened a quick draft PR to document the steps needed to deploy / have a place to discuss deployment stuff: https://github.com/Glimesh/ops/pull/41

File: src/Rtp/SequenceTracker.cpp [hunk 1]
```diff
@@ -0,0 +1,261 @@
+/**
+ * @file SequenceTracker.cpp
+ * @author Daniel Stiner (danstiner@gmail.com)
+ * @date 2021-04-17
+ * @copyright Copyright (c) 2021 Daniel Stiner
+ */
+
+#include "SequenceTracker.h"
+
+#include <spdlog/spdlog.h>
```
Comment (haydenmc) [inline src/Rtp/SequenceTracker.cpp:10]: PCH again (I think)

File: src/Rtp/SequenceTracker.cpp [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    nacks.emplace(static_cast<rtp_sequence_num_t>(extendedSeq), OutstandingNack {
+        .extendedSeq = extendedSeq,
+        .sent_at = std::chrono::steady_clock::now(),
+    });
+}
+
+std::vector<rtp_extended_sequence_num_t> SequenceTracker::GetNackList()
+{
+    // If we might exceed the maximum number of outstanding NACKs
+    if (missing.size() + nacks.size() >= MAX_OUTSTANDING_NACKS) {
```
Comment (haydenmc) [inline src/Rtp/SequenceTracker.cpp:41]: Braces on new line

File: src/Rtp/SequenceTracker.cpp [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            else
+            {
+                ++it;
+            }
+        }
+    }
+
+    // Build a list of missing packets not already NACK'd, starting with the latest missing
+    std::vector<rtp_extended_sequence_num_t> toNack;
+    for (auto it = missing.rbegin(); it != missing.rend() && toNack.size() + nacks.size() < MAX_OUTSTANDING_NACKS; ++it)
```
Comment (haydenmc) [inline src/Rtp/SequenceTracker.cpp:61]: Fold lines > 100 char pls

File: src/Rtp/SequenceTracker.cpp [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        initialized = true;
+    }
+
+    if (extendedSeq > highestReceived)
+    {
+        highestReceived = extendedSeq;
+    }
+
+    rtp_extended_sequence_num_t lowerBound = highestChecked + 1;
+    rtp_extended_sequence_num_t upperBound = highestReceived > REORDER_DELTA ? highestReceived - REORDER_DELTA : 0;
```
Comment (haydenmc) [inline src/Rtp/SequenceTracker.cpp:192]: Fold long lines

File: src/Rtp/SequenceTracker.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+#include "ExtendedSequenceCounter.h"
+#include "Types.h"
+
+#include <chrono>
+#include <list>
+#include <map>
+#include <set>
+
+using namespace std::literals;
```
Comment (haydenmc) [inline src/Rtp/SequenceTracker.h:18]: I usually avoid `using namespace` in headers, since they'll be applied to any file that #includes this file.

**Review by haydenmc — APPROVED** [review]
Approved w/ suggestions 😁

File: src/Rtp/ExtendedSequenceCounter.cpp [hunk 6]
```diff
@@ -39,77 +43,131 @@ bool ExtendedSequenceCounter::UpdateState(rtp_sequence_num_t seq)
...
-    badSeq = RTP_SEQ_MOD + 1;   /* so seq == badSeq is false */
+    badSeq = RTP_SEQ_MOD + 1; /* so seq == badSeq is false */
     cycles = 0;
-    received = 0;
     receivedPrior = 0;
     expectedPrior = 0;
 }
 
-std::ostream& operator<<(std::ostream & os, const ExtendedSequenceCounter& c)
+void ExtendedSequenceCounter::resync(rtp_sequence_num_t seq)
```
Comment (AdamTReineke) [inline src/Rtp/ExtendedSequenceCounter.cpp:161]: This naming obscures finding all calls to `reset`. Recommend removing unless you expect `resync` to diverge soon.

File: src/Rtp/ExtendedSequenceCounter.cpp [hunk 7]
```diff
@@ -8,28 +8,32 @@
 #include "ExtendedSequenceCounter.h"
 
 #include <iostream>
+#include <spdlog/spdlog.h>
```
[the remark this answers is not in the document]
Comment (danstiner) [reply src/Rtp/ExtendedSequenceCounter.cpp:11]: It is. My understanding is that best practice is for a project to still compile when pre-compiled headers are disabled. My motivation hear was that my VSCode makes lots of red squiggles without this, it seems to not use precompiled headers.

References:
- https://stackoverflow.com/a/41614105/539151
- https://mesonbuild.com/Precompiled-headers.html#toggling-the-usage-of-precompiled-headers

File: src/Rtp/ExtendedSequenceCounter.cpp (hunk 6 above)
Comment (danstiner) [reply src/Rtp/ExtendedSequenceCounter.cpp]: Fixed, thanks for the feedback!

File: src/Rtp/SequenceTracker.cpp (hunk 2 above)
Comment (danstiner) [reply src/Rtp/SequenceTracker.cpp]: 😅

File: src/Rtp/ExtendedSequenceCounter.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (haydenmc) [reply src/Rtp/ExtendedSequenceCounter.cpp:11]: Good to know - in that case at some point we should probably go and ensure the project actually compiles with PCH disabled, I suspect it won't 😅
